Implement mocknet and CSV output of HF21 transition

Doy-lee opened the pull request:
Goes ontop of https://github.com/oxen-io/oxen-core/pull/1790/files

I did have to modify the transition code a little bit to extract out information for producing the reports. The notable changes:

- Mocknet is enabled by configuring the build with `-D WITH_MOCKNET=ON`. The intention is that in our release and packagable builds everything mocknet should be completely compiled out and there's no way of getting it to activated at all in public builds.

Most of the implementation is compartmentalised into `src/network_config/mocknet.{h,cpp}` and compiles to no-op macros when disabled.

- Some cursory header cleanup, removing unused headers, moving things to implementation file if it can be, removing dead/unused macros e.t.c.

- Updating the conversion_ratio to use 64 bits, the numbers we were playing were fractional and we have a `mul128_div` primitive in the codebase that seemed fine to use for this purpose.

- Stack trace support is not a build flag enabled with `-D WITH_STACKTRACE=ON` 

- Fix 0 OXEN payout generating an output in the miner TX on transition of the amount of rewards owed is less than 1000 DB OXEN units.

Enabling mocknet can be done by specifying the flag `--fork-to-mocknet <height>` at which when the chain height reaches `<height>` (e.g. `top_block.get_height() + 1 == <height>`) then mock nodes will be injected into the SNL and all the quorums from that point will be hijacked by those nodes and it will periodically generate blocks to the chain.

When this happens, pulse timestamps are disabled and blocks are appended to the chain every 5s.

jagerman committed:
WIP oxen->sent transition code

tewinget committed:
HF21 transition code mostly finished

the bulk of the code for transitioning to HF21 is present, and much
testing code is present.  This code does not yet work, and the tests are
not yet complete.

add hf21 transition mechanism to sqlite db

tewinget committed:
Post-hf21 purge and deactivation limit

After HF21, any nodes which did not transition to SENT stakes need to be
removed.  Doing them all at once would be chaotic with swarms, so they
will be removed by the existing decommission/deregister system.

That system now has a config-determined limit of deactivations per block
so that this transition (but also huge network outages) won't cause
swarms to disappear (or other possible chaos resulting from mass
removal).

tewinget committed:
Final HF20 block payout

Staked oxen addresses which are not registered for the SENT transition
will have their pending service node rewards paid out in full on the
final block before HF21.

Modified local devnet testing to leave a staked static-startup address
out of the registration for testing this (and adjusted expected numbers
accordingly).

tewinget committed:
correct removal of nodes which did not go past hf21

nodes which did not transition to hf21 will be removed by the
decommission->deregister mechanism rather than by the
bls-key-not-in-the-contract mechanism

tewinget committed:
HF21 fixes and debug logging

tewinget committed:
rough-but-complete script for hf21 transition info

script needs parameters set as appropriate:
  - oxen -> eth address mapping
  - transition bonus mapping
  - staking requirement and scaling ratio

TODO: some mainnet nodes have a larger stake, scaling needs to be normalized for those

tewinget committed:
change localdev initial staking req to match contract

tewinget committed:
testnet hf21 height

tewinget committed:
change testnet netid for chain restart

tewinget committed:
Don't apply HF dereg grace period to zombie nodes

They cannot ever come back, so there's really no point of leaving them
for a week.

(We *don't* want to immediately dereg them because going into decomm
still allows SN comms between storage servers, allowing the leaving
nodes to still transfer their files out to the remaining nodes upon
swarm dissolution).

tewinget committed:
make hf21 zombies not payable

nodes which do not transition at hardfork 21 should not be
in the reward payment rotation

tewinget committed:
remove erroneous hf21 transition change

tewinget committed:
add hf21 dry run rpc method

calling this method returns a before and after as if hf21 would
happen the very next block.  Because outstanding oxen rewards will
differ from the exact fork height (and from block to block), this
can only be approximate, but it should be fairly close to what the
actual result will be.

tewinget committed:
local devnet retry on json-rpc 'error'

tewinget committed:
hf21 transition: always set bls key if present, even if node fails transition

tewinget committed:
devnet restart for hf21 testing

tewinget committed:
add devnet reset old block hashes

tewinget committed:
devnet temp hf heights, double staking req

higher staking requirement to test hf21 transition correctly scaling stake
proportions in multi-contrib

tewinget committed:
fix inverted condition in daemon command

tewinget committed:
devnet timings underflow fix (and slow it down a little)

tewinget committed:
devnet hf20 height

tewinget committed:
(reverted later) devnet hf21 height and stop chain before

one line of this commit was kept and moved, making actually reverting it
and demolishing this commit tedious, so it stays, but the important
aspect of it which was to be reverted was later reverted.

tewinget committed:
redo/fix pre-hf21 bulk payout for non-registrants

The block before HF21 all addresses with accrued rewards which are
not registered for the HF21 transition with an ETH address will have
their OXEN balance paid out at that height rather than converted.

This was not working correctly for multiple reasons, but is now.

tewinget committed:
push devnet hf21 back a little

tewinget committed:
devnet hf21 parameters

tewinget committed:
fix uptime proofs using hf rules one block too early

tewinget committed:
stop using old monero-style keys altogether at HF21

tewinget committed:
fix stake scaling for hf21 transition for over-funded nodes

tewinget committed:
fix hf21_info stake scaling for over-funded nodes

also update devnet hf21 parameters

tewinget committed:
change devnet l2 tracker parameters

tewinget committed:
remove devnet stall-before-hf21 code

tewinget committed:
devnet hf dereg grace period 100 blocks

tewinget committed:
devnet reboot for testing mainnet test contracts

tewinget committed:
cleanup

tewinget committed:
clang format

Doy-lee committed:
Move network config args into command_line.cpp

Doy-lee committed:
Move network config args into command_line.cpp

Doy-lee committed:
Remove unused macro BINARY_ARCHIVE_STREAM

tewinget committed:
hf21 ensure 'zombies' do not linger in recently removed

nodes which fail hf21 transition but were seeded into the contract need
to be removeable from the contract.  the ones which were *not* seeded
into the contract would not be removable from `recently_removed_nodes`
so should not be added there to begin with

Doy-lee committed:
Optimize/remove unused headers in command_line.h

- Windows.h is only needed if we don't have readline
- Use WIN32_LEAN_AND_MEAN to #define out a bunch of headers you don't need

Doy-lee committed:
Remove unused headers from core

Doy-lee committed:
Rmove unused offline field in Blockchain

Doy-lee committed:
Add mocknet support via --fork-to-mocknet

Doy-lee committed:
Remove unused headers in blockchain

Doy-lee committed:
Make stacktrace a cmake build option, WITH_STACKTRACE

Doy-lee committed:
Enable mocknet through CMake option WITH_MOCKNET

Doy-lee committed:
Mocknet changes to work on mainnet

- Address some issues with the network not correctly transitioning to
mocknet when a height is given that has already passed.
- Fix block leader not being set in the cache in certain scenarios
causing bad blocks to be generated
- Ensure that the Pulse quorum is hijacked at the leaf-most function for
generating a quorum to ensure all code paths requesting a Pulse quorum correctly
get hijacked with mock nodes
- Simplify the mocknet interface by dropping the post CN block add hook
and instead kill P2P comms when a mock pulse block is being constructed

Doy-lee committed:
Gen data for making transition CSVs, block purging in mocknet

tewinget committed:
boost 1.87 deprecated removal fixes

Boost 1.87 removed many things which were previously only deprecated,
some of which we were still using.

tewinget committed:
remove boost version guard on typedef

changing from including boost asio io_service to io_context is roughly equivalent,
but the former has a typedef for compatibility.  I replicated the typedef,
but gated it on the boost version that removed the io_service include.

tewinget committed:
bump static build boost to 1.87

tewinget committed:
clang format to appease the lint gods

tewinget committed:
undo superfluous (invalid) boost-related change

tewinget committed:
boost is so much fun

tewinget committed:
thank you again, clang format!

Doy-lee committed:
Final HF21 payments must not payout 0 OXEN when truncated from DB OXEN

Doy-lee committed:
Allow money printing to take custom decimal places

Doy-lee committed:
Add more detail to the CSV dumps of the token transition

Doy-lee committed:
Hide mocknet impl, pull transition parameters into struct

Doy-lee committed:
Pass the transition context into the transition function

This lets the caller control the final values that the transition will run with.
For ordinary usecases this is not important, the transition parameters
are pulled from the compile time constants. For testing and mocking it
can be useful to modify these values at runtime to simulate different
transition conditions.

Doy-lee committed:
Remove unused parse_oxen/eth_address from sent_transition.h

Doy-lee committed:
Calc transition STAKING_RATIO at runtime, make conversion ratio 64bits

For mocknet and testing we now allow changing the staking requirement
at runtime (i.e. for testing different staking requirements and
conversion ratios) the staking ratio needs to be updated with the
appropriate values.

Allow a higher precision conversion ratio to allow for more fine grained
conversions. For the $OXEN to $SESH conversion we now defer to a 128 bit
multiply to avoid overflow for the conversion.

Doy-lee committed:
Simplify mocknet construction of the transition parameters

Doy-lee committed:
Revert some incorrect transition stat calculations

I was counting the number SESH calculated for each contributor as the
amount they contributed to the node. This was incorrect because that
number is not the amount they were able to stake, just the amount the
were required to have to be able to keep the node alive.

This mean that `tokens_allocated >= staking_requirement` was incorrectly
attributing a node as having made it past the transition when it
actually didn't.

Instead of trying to be clever and calculating the partial collaterised
nodes that will be zombified, only count tokens from a fully
collateralised node.

Doy-lee committed:
Add sanity check of HF transition code when producing CSV reports

Doy-lee committed:
Add option to dump the transition CSVs, fix incorrect mocknet conversion ratio

Doy-lee committed:
SESH rewards need to be stored in DB units

Doy-lee committed:
Remove debugging hardforks for testing mocknet

Doy-lee committed:
Adjust mocknet to fork at chain height to follow codebase convention

Doy-lee committed:
Add braces around no-op macro when mocknet is disabled

Doy-lee committed:
Fix mocknet comment, use oxen::array_count

Doy-lee committed:
Linting pass

Doy-lee committed:
Avoid querying block height before miner_tx is instantiated

tewinget commented on src/blockchain_db/sqlite/db_sqlite.cpp:
```diff
@@ -1102,4 +1144,29 @@ bool BlockchainSQLite::save_payments(
+
+    for (auto& [addr, amt] : rewards) {
+        auto address_str = "0x{:x}"_format(addr);
+        auto amount = static_cast<int64_t>(amt * BATCH_REWARD_FACTOR);
```
Should be covered in #1802 but github won't let me comment here in that PR.  I assume that this `amt * BATCH_REWARD_FACTOR` should be converted to the "db vs actual" coin functions?

Doy-lee replied:
Yep. Just separated them because that change could be completely isolated out and that this PR focuses on just getting mocknet in.

tewinget commented on src/cryptonote_core/cryptonote_core.cpp:
```diff
@@ -454,6 +451,8 @@ bool core::handle_command_line(const boost::program_options::variables_map& vm)
         }
     }
 
+    if (!mocknet_read_cli_for_mocknet_arg(vm, m_service_node))
```
tiniest nitpick ever, perhaps, but simply calling this function doesn't necessarily (at least in this codebase) imply that "mocknet" is fundamentally required condition here.  I imagine future-me could read this and wonder what that will return, and assume (but not know for sure) that "mocknet" is false.

tewinget replied:
On second glance, I was just overthinking.

tewinget reviewed (approved):
Couple tiny nitpicks but LGTM

Doy-lee committed:
Use constexpr bitset size in bytes to avoid clang-format bug

```diff
     std::bitset<sizeof(uint16_t) * 8> bitset_view16(uint16_t val) {
-        std::bitset<sizeof(uint16_t) * 8> result = val;
+        std::bitset<sizeof(uint16_t)* 8> result = val;
         return result;
     }
```

clang-format constantly tries to re-format this line, moving the '*'
back and forth all the time. Seems to be a bug that we can avoid
completely by just pulling out the expression into a constexpr value.

tewinget committed:
fix sent transition circular dependency

Doy-lee committed:
Fix build w/ sent_transition moved to cryptonote_core
